### PR TITLE
[FIX] gcc-11: satisfaction of atomic constraint depends on itself

### PR DESCRIPTION
**Pull request by marehr**
Fixes https://github.com/seqan/product_backlog/issues/371

### CODE REVIEW

**Comment by eseiler** [comment]
Interesting, I could "reduce" it to

```cpp
#include <seqan3/alphabet/adaptation/char.hpp>
#include <seqan3/alphabet/gap/gapped.hpp>
#include <seqan3/alphabet/nucleotide/dna4.hpp>
#include <seqan3/range/container/bitcompressed_vector.hpp>

int main()
{
   using alphabet_t = seqan3::alphabet_variant<char, seqan3::dna4>;
   seqan3::bitcompressed_vector<alphabet_t> source{};
   auto it = source.begin();
   auto end = source.end();
   it != end; // This line causes error.
}
```

And could fix is making the comparison operators friends
```patch
diff --git a/include/seqan3/range/container/bitcompressed_vector.hpp b/include/seqan3/range/container/bitcompressed_vector.hpp
index baddc729..44b9ea62 100644
--- a/include/seqan3/range/container/bitcompressed_vector.hpp
+++ b/include/seqan3/range/container/bitcompressed_vector.hpp
@@ -1077,18 +1077,18 @@ public:
      * \details
      * \stableapi{Since version 3.1.}
      */
-    constexpr bool operator==(bitcompressed_vector const & rhs) const noexcept
+    friend constexpr bool operator==(bitcompressed_vector && lhs, bitcompressed_vector && rhs) noexcept
     {
-        return data == rhs.data;
+        return lhs.data == rhs.data;
     }
 
     /*!\brief Checks whether `*this` is not equal to `rhs`.
      * \details
      * \stableapi{Since version 3.1.}
      */
-    constexpr bool operator!=(bitcompressed_vector const & rhs) const noexcept
+    friend constexpr bool operator!=(bitcompressed_vector && lhs, bitcompressed_vector && rhs) noexcept
     {
-        return data != rhs.data;
+        return lhs.data != rhs.data;
     }
 
     /*!\brief Checks whether `*this` is less than `rhs`.

```

**Comment by marehr** [comment]
> I could "reduce" it to
> 
> ```c++
> #include <seqan3/alphabet/adaptation/char.hpp>
> #include <seqan3/alphabet/gap/gapped.hpp>
> #include <seqan3/alphabet/nucleotide/dna4.hpp>
> #include <seqan3/range/container/bitcompressed_vector.hpp>
> 
> int main()
> {
>    using alphabet_t = seqan3::alphabet_variant<char, seqan3::dna4>;
>    seqan3::bitcompressed_vector<alphabet_t> source{};
>    auto it = source.begin();
>    auto end = source.end();
>    it != end; // This line causes error.
> }
> ```
> 
> And could fix it by making the comparison operators friends
> 
> ```diff
> diff --git a/include/seqan3/range/container/bitcompressed_vector.hpp b/include/seqan3/range/container/bitcompressed_vector.hpp
> index baddc729..44b9ea62 100644
> --- a/include/seqan3/range/container/bitcompressed_vector.hpp
> +++ b/include/seqan3/range/container/bitcompressed_vector.hpp
> @@ -1077,18 +1077,18 @@ public:
>       * \details
>       * \stableapi{Since version 3.1.}
>       */
> -    constexpr bool operator==(bitcompressed_vector const & rhs) const noexcept
> +    friend constexpr bool operator==(bitcompressed_vector && lhs, bitcompressed_vector && rhs) noexcept
>      {
> -        return data == rhs.data;
> +        return lhs.data == rhs.data;
>      }
>  
>      /*!\brief Checks whether `*this` is not equal to `rhs`.
>       * \details
>       * \stableapi{Since version 3.1.}
>       */
> -    constexpr bool operator!=(bitcompressed_vector const & rhs) const noexcept
> +    friend constexpr bool operator!=(bitcompressed_vector && lhs, bitcompressed_vector && rhs) noexcept
>      {
> -        return data != rhs.data;
> +        return lhs.data != rhs.data;
>      }
>  
>      /*!\brief Checks whether `*this` is less than `rhs`.
> ```

Nice, I set on this forever xD

Rock, paper, scissor?

(My fix is more general, as it excludes the potential of recursion which I also had trouble with clang, your test case is more simple and shows the actual problem xD)

**Comment by eseiler** [comment]
> Nice, I sat on this forever xD

I had the right hunch so it just took a few minutes (for once)

> Rock, paper, scissor?
>
> (My fix is more general, as it excludes the potential of recursion which I also had trouble with clang, your test case is more simple and shows the actual problem xD)

I would definitely change the testing example, because I think `it != end` is clearer.

The range-based for loops does the same (https://cppinsights.io/s/d1d6f5e9)
But the `c = a` indicates that this may be a problem with assignment which it is not.

Restricting `other_range_t` to be not recursive seems also correct.
IIRC we want to have comparison operators as friends, no?

Then maybe we just do both changes?

FYI gcc11 introduces new template matching: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0522r0.html
Haven't read it yet but maybe related

**Comment by marehr** [comment]
> FYI gcc11 introduces new template matching: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0522r0.html
> Haven't read it yet but maybe related

Nice, I didn't know that there was a paper for this. This bothered my a couple of times already :)

**Comment by marehr** [comment]
> IIRC we want to have comparison operators as friends, no?

I personally like them to be members, especially since C++20 where also the commutative ones are auto-generated.

But, let's just apply that and see if anything breaks :)

**Comment by eseiler** [comment]
> > IIRC we want to have comparison operators as friends, no?
> 
> I personally like them to be members, especially since C++20 where also the commutative ones are auto-generated.
> 
> But, let's just apply that and see if anything breaks :)

I am also fine with just doing the `non-recursive` thing.

File: test/unit/range/container/bitcompressed_vector_test.cpp [hunk 1]
```diff
@@ -39,3 +41,13 @@ TEST(bitcompressed_vector_test, issue1743_view_combinability)
...
 }
+
+TEST(bitcompressed_vector_test, issue371)
+{ // https://github.com/seqan/product_backlog/issues/371
+    using alphabet_t = seqan3::alphabet_variant<seqan3::dna4, seqan3::dna15>;
+    seqan3::bitcompressed_vector<alphabet_t> source{};
+    alphabet_t a{};
+
+    for (auto && c : source)
+        c = a;
```
Comment (marehr) [inline test/unit/range/container/bitcompressed_vector_test.cpp:52]:
```suggestion
    using alphabet_t = seqan3::alphabet_variant<seqan3::dna4, seqan3::dna15>;
    seqan3::bitcompressed_vector<alphabet_t> source{};
    auto it = source.begin();
    auto end = source.end();
    it != end; // This line causes error.
```

File: include/seqan3/range/container/bitcompressed_vector.hpp [hunk 2]
```diff
@@ -1077,18 +1079,18 @@ class bitcompressed_vector
...
-        return data == rhs.data;
+        return lhs.data == rhs.data;
     }
 
     /*!\brief Checks whether `*this` is not equal to `rhs`.
      * \details
      * \stableapi{Since version 3.1.}
      */
-    constexpr bool operator!=(bitcompressed_vector const & rhs) const noexcept
+    friend constexpr bool operator==(bitcompressed_vector && lhs, bitcompressed_vector && rhs) noexcept
```
Comment (eseiler) [inline include/seqan3/range/container/bitcompressed_vector.hpp:1091]:
```suggestion
    friend constexpr bool operator!=(bitcompressed_vector && lhs, bitcompressed_vector && rhs) noexcept
```

File: include/seqan3/range/container/bitcompressed_vector.hpp [hunk 3]
```diff
@@ -1077,18 +1079,18 @@ class bitcompressed_vector
      * \details
      * \stableapi{Since version 3.1.}
      */
-    constexpr bool operator==(bitcompressed_vector const & rhs) const noexcept
+    friend constexpr bool operator==(bitcompressed_vector && lhs, bitcompressed_vector && rhs) noexcept
```
Comment (marehr) [inline include/seqan3/range/container/bitcompressed_vector.hpp:1082]:
```suggestion
        constexpr bool operator==(bitcompressed_vector const & rhs) const noexcept
```

File: include/seqan3/range/container/bitcompressed_vector.hpp [hunk 4]
```diff
@@ -1077,18 +1079,18 @@ class bitcompressed_vector
...
     {
-        return data == rhs.data;
+        return lhs.data == rhs.data;
```
Comment (marehr) [inline include/seqan3/range/container/bitcompressed_vector.hpp:1084]:
```suggestion
        return data == rhs.data;
```

File: include/seqan3/range/container/bitcompressed_vector.hpp [hunk 5]
```diff
@@ -1077,18 +1079,18 @@ class bitcompressed_vector
...
     }
 
     /*!\brief Checks whether `*this` is not equal to `rhs`.
      * \details
      * \stableapi{Since version 3.1.}
      */
-    constexpr bool operator!=(bitcompressed_vector const & rhs) const noexcept
+    friend constexpr bool operator!=(bitcompressed_vector && lhs, bitcompressed_vector && rhs) noexcept
```
Comment (marehr) [inline include/seqan3/range/container/bitcompressed_vector.hpp:1091]:
```suggestion
    constexpr bool operator!=(bitcompressed_vector const & rhs) const noexcept
```

File: include/seqan3/range/container/bitcompressed_vector.hpp [hunk 6]
```diff
@@ -1077,18 +1079,18 @@ class bitcompressed_vector
...
     {
-        return data != rhs.data;
+        return lhs.data != rhs.data;
```
Comment (marehr) [inline include/seqan3/range/container/bitcompressed_vector.hpp:1093]:
```suggestion
        return data != rhs.data;
```

File: include/seqan3/range/container/bitcompressed_vector.hpp [hunk 7]
```diff
@@ -1077,7 +1079,7 @@ class bitcompressed_vector
      * \details
      * \stableapi{Since version 3.1.}
      */
-    constexpr bool operator==(bitcompressed_vector const & rhs) const noexcept
+        constexpr bool operator==(bitcompressed_vector const & rhs) const noexcept
```
Comment (marehr) [inline include/seqan3/range/container/bitcompressed_vector.hpp:1082]:
```suggestion
    constexpr bool operator==(bitcompressed_vector const & rhs) const noexcept
```

File: test/unit/range/container/bitcompressed_vector_test.cpp [hunk 8]
```diff
@@ -39,3 +41,12 @@ TEST(bitcompressed_vector_test, issue1743_view_combinability)
...
 }
+
+TEST(bitcompressed_vector_test, issue371)
+{ // https://github.com/seqan/product_backlog/issues/371
+    using alphabet_t = seqan3::alphabet_variant<seqan3::dna4, seqan3::dna15>;
+    seqan3::bitcompressed_vector<alphabet_t> source{};
+    auto it = source.begin();
+    auto end = source.end();
+    it != end; // This line causes error.
+}
```
Comment (eseiler) [inline test/unit/range/container/bitcompressed_vector_test.cpp:52]:
```suggestion

// https://github.com/seqan/product_backlog/issues/371
TEST(bitcompressed_vector_test, issue371)
{ 
    using alphabet_t = seqan3::alphabet_variant<seqan3::dna4, seqan3::dna15>;
    seqan3::bitcompressed_vector<alphabet_t> source{};
    auto it = source.begin();
    auto end = source.end();
    it != end; // This line causes error.
}
```

**Review by MitraDarja — APPROVED** [review]

**Review by eseiler — APPROVED** [review]

**Pull request merged**